internal/lib/container_server.go

```diff
@@ -272,40 +271,76 @@ func (c *ContainerServer) LoadSandbox(id string) (retErr error) {
 		}
 	}()
 
-	scontainer, err := oci.NewContainer(m.Annotations[annotations.ContainerID], cname, sandboxPath, m.Annotations[annotations.LogPath], labels, m.Annotations, kubeAnnotations, m.Annotations[annotations.Image], "", "", nil, id, false, false, false, sb.RuntimeHandler(), sandboxDir, created, m.Annotations["org.opencontainers.image.stopSignal"])
-	if err != nil {
-		return err
+	var scontainer *oci.Container
+
+	// we should not take whether the server actively has DropInfra specified, but rather
+	// whether the server used to.
+	wasSpoofed := false
+	if spoofed, ok := m.Annotations[crioann.SpoofedContainer]; ok && spoofed == "true" {
+		wasSpoofed = true
 	}
-	scontainer.SetSpec(&m)
-	scontainer.SetMountPoint(m.Annotations[annotations.MountPoint])
 
-	if m.Annotations[annotations.Volumes] != "" {
-		containerVolumes := []oci.ContainerVolume{}
-		if err = json.Unmarshal([]byte(m.Annotations[annotations.Volumes]), &containerVolumes); err != nil {
-			return fmt.Errorf("failed to unmarshal container volumes: %v", err)
+	if !wasSpoofed {
+		scontainer, err = oci.NewContainer(m.Annotations[annotations.ContainerID], cname, sandboxPath, m.Annotations[annotations.LogPath], labels, m.Annotations, kubeAnnotations, m.Annotations[annotations.Image], "", "", nil, id, false, false, false, sb.RuntimeHandler(), sandboxDir, created, m.Annotations["org.opencontainers.image.stopSignal"])
+		if err != nil {
+			return err
+		}
+		scontainer.SetSpec(&m)
+		scontainer.SetMountPoint(m.Annotations[annotations.MountPoint])
+
+		if m.Annotations[annotations.Volumes] != "" {
+			containerVolumes := []oci.ContainerVolume{}
+			if err = json.Unmarshal([]byte(m.Annotations[annotations.Volumes]), &containerVolumes); err != nil {
+				return fmt.Errorf("failed to unmarshal container volumes: %v", err)
+			}
+			for _, cv := range containerVolumes {
+				scontainer.AddVolume(cv)
+			}
+		}
+
+		if err := c.ContainerStateFromDisk(scontainer); err != nil {
+			return fmt.Errorf("error reading sandbox state from disk %q: %v", scontainer.ID(), err)
 		}
-		for _, cv := range containerVolumes {
-			scontainer.AddVolume(cv)
+
+		// We write back the state because it is possible that crio did not have a chance to
+		// read the exit file and persist exit code into the state on reboot.
+		if err := c.ContainerStateToDisk(scontainer); err != nil {
+			return fmt.Errorf("failed to write container state to disk %q: %v", scontainer.ID(), err)
```

Comment:
maybe put container ID after the word "container"? As in `failed to write container %q state to disk...`